core/internal/filetransfer/file_transfer_gcs.go

```diff
@@ -0,0 +1,243 @@
+	if task.Reference == nil {
+		err := fmt.Errorf("gcs file transfer: download: reference is nil")
+		ft.logger.CaptureError(err)
+		return err
+	}
+	reference := *task.Reference
+
+	// Parse the reference path to get the scheme, bucket, and object
+	uriParts, err := url.Parse(reference)
+	if err != nil {
+		ft.logger.CaptureError(fmt.Errorf("gcs file transfer: download: error parsing reference %s: %v", reference, err))
+		return err
+	}
+	if uriParts.Scheme != "gs" {
+		err := fmt.Errorf("gcs file transfer: download: invalid gsutil URI %s", reference)
+		ft.logger.CaptureError(err)
+		return err
+	}
+	bucketName := uriParts.Host
+	bucket := ft.client.Bucket(bucketName)
+	objectName := strings.TrimPrefix(uriParts.Path, "/")
+
+	switch {
+	case task.Digest == reference:
+		// If not using checksum, get all objects under the reference
+		var objectNames []string
+		query := &storage.Query{Prefix: objectName}
+		it := bucket.Objects(ft.ctx, query)
+		for {
+			objAttrs, err := it.Next()
+			if err == iterator.Done {
+				break
+			}
+			if err != nil {
+				ft.logger.CaptureError(fmt.Errorf("gcs file transfer: download: error iterating through gcs bucket %s with prefix %s for reference %s: %v", bucketName, objectName, reference, err))
+				return err
+			}
+			objectNames = append(objectNames, objAttrs.Name)
+		}
+
+		g := new(errgroup.Group)
+		g.SetLimit(maxWorkers)
+		for _, name := range objectNames {
+			objName := name // for closure in the goroutine
+			g.Go(func() error {
+				object := bucket.Object(objName)
+				localPath := getDownloadFilePath(objName, objectName, task.Path)
+				err := ft.DownloadFile(object, localPath)
+				if err != nil {
+					// this file is probably a folder, skip to avoid an unnecessary error
+					if isDir(reference, task.Size) {
+						ft.logger.Debug("gcs file transfer: download: skipping reference because it seems to be a folder", "reference", reference)
+						return nil
+					}
+					ft.logger.CaptureError(fmt.Errorf("gcs file transfer: download: error when downloading file for reference %s: %v", reference, err))
+					return err
+				}
+				return nil
+			})
+		}
+
+		// Wait for all of the go routines to complete and return an error if any errored out
+		if err := g.Wait(); err != nil {
+			return err
+		}
+	case task.VersionId != nil:
+		object := bucket.Object(objectName).Generation(int64(task.VersionId.(float64)))
+		err := ft.DownloadFile(object, task.Path)
+		if err != nil {
+			// this file is probably a folder, skip to avoid an unnecessary error
+			if isDir(reference, task.Size) {
+				ft.logger.Debug("gcs file transfer: download: skipping reference because it seems to be a folder", "reference", reference)
+				return nil
+			}
+			ft.logger.CaptureError(fmt.Errorf("gcs file transfer: download: error while downloading file for reference %s: %v", reference, err))
+			return err
+		}
+	default:
```

Comment:
from what i could tell, our gcs manifest entries always have versionId populated. if that's something we can say for sure then maybe I should just remove this case